### PR TITLE
Reorganize route structure

### PR DESCRIPTION
**Pull request by tusbar**
**This PR introduces a new routing structure:**

- Routes are now nested under `src/routes`.
- They are defined using POJO instead of JSX: this allows us to create code split points and serve the different routes in different bundles.

**Some of the routes have NOT been migrated yet, and will be in a future PR:**

- `src/routes/Datasets` (components are still under `src/modules/Datasets`)
- `src/routes/Publication` (components are still under `src/modules/Publication`)

**Some heavy modules have been removed:**

- `qs` in favor of the (almost) native `querystring` module.
- `lodash.isequal` (used by react-chartjs-2) is now aliased as `lodash/isEqual` in order to remove duplicated code from the bundle.

**Some other changes are also introduced in this PR:**

- Redux: a global redux store has been added to simplify data fetching and components. Most of the components that have been migrated are now functional components, or pure.
- Webpack has been updated to 3.3.0 for build performance purposes.
- `src/` has been aliased to `common/` in both webpack and test environments, this allows referencing shared components as `common/components/Loader`, for example, instead of `../../../../../../components/Loader`.
- More configuration variables have been created to remove some of the hardcoded API URLs.
- A few visual changes have also been added to this PR as some of the CSS was duplicated in different places, or was ugly. :)
- PropTypes, PropTypes and PropTypes. This helps. A lot. Eslint errors and warnings are coming soon. 😈 

**The migrated routes are:**

```
/catalogs
├─ /catalogs/:id
├─ /catalogs/:id/harvest/:id
/events
/home
/search
```

And the `*` (not found) route.

**Overview of the updated bundles:**

```
File sizes after gzip:

  dist/scripts/vendor.js            164.47 KB
  dist/scripts/app.js               32.99 KB
  dist/styles/app.css               8.63 KB
  dist/scripts/modules/catalogs.js  6 KB
  dist/scripts/modules/search.js    5.14 KB
  dist/index.html                   1.27 KB
  dist/scripts/manifest.js          942 B
  dist/scripts/modules/events.js    869 B

  TOTAL                             220.27 KB
```

https://1171-67130793-gh.circle-artifacts.com/0/home/ubuntu/inspire/reports/bundles.html

The overall size is a little more than on master (by 10KB) for a few reasons:
- redux (1.76 KB)
- react-redux (3.54 KB)
- Some overhead by splitting the modules.
- Some CSS was added.

Removing the last 2 modules (datasets & publication) will greatly reduce the initial load time by reducing `app.js` bundle size.

**Linked issues**

Fixes partially #360, will fix completely once Dataset and Publication are migrated.

### CODE REVIEW

**Comment by jdesboeufs** [comment]
C'est encore WIP ?
Le jeu. 20 juil. 2017 à 17:43, Bertrand Marron <notifications@github.com> a
écrit :

**Comment by tusbar** [comment]
@jdesboeufs ouais je finis la description, et je vais push un patch pour pas autodeploy master avant (#410).

**Comment by tusbar** [comment]
I have some issues with the production build as some stuff breaks, for some reason. Will investigate.

**Comment by tusbar** [comment]
@jdesboeufs maintenant c'est 👍

File: src/routes/Catalogs/routes/Catalog/components/CatalogView/CatalogView.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            <CatalogHarvestsView
+              catalog={catalog}
+              harvests={harvests}
+              syncCatalog={syncCatalog}
+            />
+          </div>
+
+          <div className={styles.section}>
+            <h2>Rechercher dans les jeux de données du catalogue</h2>
+            <SearchInput onSearch={this.onSearch} hasButton />
```
Comment (tmerlier) [inline src/routes/Catalogs/routes/Catalog/components/CatalogView/CatalogView.js:78]: J'ai toujours été obligé de faire `hasButton={hasButton}`, quelque de nouveau permet de le faire ou est-ce une erreur ?

Comment (tusbar) [reply src/routes/Catalogs/routes/Catalog/components/CatalogView/CatalogView.js:78]: Y a pas de variable `hasButton` ici, cela correspond à `hasButton={true}`.

Comment (tusbar) [reply src/routes/Catalogs/routes/Catalog/components/CatalogView/CatalogView.js:78]: => https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-boolean-value.md

**Review by tmerlier — APPROVED** [review]

**Pull request merged**